qiita_ware/ebi.py

```diff
@@ -143,33 +126,76 @@ def __init__(self, preprocessed_data_id, action):
+        self.experiment_xml_fp = get_output_fp('experiment.xml')
+        self.run_xml_fp = get_output_fp('run.xml')
+        self.submission_xml_fp = get_output_fp('submission.xml')
         self.pmids = s.pmids
 
-        # dicts that map investigation_type to library attributes
-        lib_strategies = {'metagenome': 'POOLCLONE',
-                          'mimarks-survey': 'AMPLICON'}
-        lib_selections = {'mimarks-survey': 'PCR'}
-        lib_sources = {}
-
-        self.library_strategy = lib_strategies.get(key, "OTHER")
-        self.library_source = lib_sources.get(key, "METAGENOMIC")
-        self.library_selection = lib_selections.get(key, "unspecified")
-
-        # This will hold the submission's samples, keyed by the sample name
+        # ignore data type for the restrictions as we assume that this was
+        # validated in the prep template, here we only going to validate that
+        # they exists
+        # getting the restrictions
+        restrictions = [v.columns.keys()
+                        for _, v in viewitems(SAMPLE_TEMPLATE_COLUMNS)]
+        restrictions.extend([v.columns.keys()
+                             for _, v in viewitems(PREP_TEMPLATE_COLUMNS)])
+        if self.prep_template.data_type() in TARGET_GENE_DATA_TYPES:
+            restrictions.extend([v.columns.keys() for _, v in
+                                 viewitems(PREP_TEMPLATE_COLUMNS_TARGET_GENE)])
+        restrictions = set(flatten(restrictions))
+        # getting the existing columns
+        templates_categories = self.sample_template.categories()
+        templates_categories.extend(self.prep_template.categories())
+        existing_columms = set(templates_categories)
+        missing_columns = restrictions.difference(existing_columms)
+        # testing if there are any missing columns
+        if missing_columns:
+            error_msgs.append(
+                "You are missing some columns in study #%d to have a valid "
+                "submission #%d: %s." % (s.id, preprocessed_data_id,
+                                         ', '.join(list(missing_columns))))
+
+        # generating all samples from sample template
+        # self.samples = dict(self.sample_template.items())
         self.samples = {}
+        self.samples_prep = {}
+        nvp = []
+        for k, v in self.sample_template.items():
+            if k not in self.prep_template:
+                continue
+            sample_prep = self.prep_template[k]
+
+            # validating required fields
+            if 'platform' not in sample_prep:
+                nvp.append(k)
+            else:
+                platform = sample_prep['platform'].upper()
+                if platform not in self.valid_platforms:
+                    nvp.append(k)
+
+            self.samples[k] = v
+            self.samples_prep[k] = self.prep_template[k]
+
+        if nvp:
+            error_msgs.append(
+                "These samples from study #%d and preprocessed data #%d do "
```

Comment:
I _think_ it might be more useful to provide the prep template id, as the prep template is the one who should include the correct value for platform